src/analyses/ai_history.h

```diff
@@ -0,0 +1,237 @@
+  {
+  }
+
+  enum step_status
+  {
+    NEW,
+    MERGED,
+    BLOCKED
+  };
+
+  /// In a number of places we need to work with sets of histories
+  /// so that these (a) make use of the immutability and sharing and
+  /// (b) ownership can be transfered if necessary, we use shared pointers
+  /// rather than references.
+  /// The custom comparison allows particular histories to control
+  /// the order of exploration.
+  struct compare_historyt
+  {
+    bool operator()(const trace_ptrt &l, const trace_ptrt &r) const
+    {
+      return *l < *r;
+    }
+  };
+  typedef std::set<trace_ptrt, compare_historyt> trace_sett;
+
+  /// Step creates a new history by advancing the current one to location "to"
+  /// It is given the set of all other histories that have reached this point.
+  ///
+  /// PRECONDITION(to.id_dereferenceable());
+  /// PRECONDITION(to in goto_program.get_successors(current_location()) ||
+  ///              current_location()->is_function_call() ||
+  ///              current_location()->is_end_function());
+  ///
+  /// Step may do one of three things :
+  ///  1. Create a new history object and return a pointer to it
+  ///     This will force the analyser to continue regardless of domain changes
+  ///     POSTCONDITION(IMPLIES(result.first == NEW,
+  ///                           result.second.use_count() == 1 ));
+  ///  2. Merge with an existing history
+  ///     This will mean the analyser will continue of the domain is modified
+  ///     POSTCONDITION(IMPLIES(result.first == MERGED,
+  ///                           result.second is an element of others));
+  ///  3. Block this flow of execution
+  ///     This indicates the transition is impossible (returning to a location
+  ///     other than the call location) or undesireable (omitting some traces)
+  ///     POSTCONDITION(IMPLIES(result.first == BLOCKED,
+  ///                           result.second == nullptr()));
+  typedef std::pair<step_status, trace_ptrt> step_returnt;
+  virtual step_returnt step(locationt to, const trace_sett &others) const = 0;
+
+  /// The order for history_sett
+  virtual bool operator<(const ai_history_baset &op) const = 0;
+
+  /// History objects should be comparable
+  virtual bool operator==(const ai_history_baset &op) const = 0;
+
+  /// The current location in the history, used to compute the transformer
+  /// POSTCONDITION(return value is dereferenceable)
+  virtual const locationt &current_location(void) const = 0;
+
+  /// Domains with a substantial height may need to widen when merging
+  /// this method allows the history to provide a hint on when to do this
+  /// The suggested use of this is for domains merge operation to check this
+  /// and then widen in a domain specific way.  However it could be used in
+  /// other ways (such as the transformer).  The key idea is that the history
+  /// tracks / should know when to widen and when to continue.
+  virtual bool widen(const ai_history_baset &other) const
```

Comment:
If this is just a predicate, then maybe it could be called something like `should_widen`. Just `widen` sounds like an imperative.

Reply:
OK; `should_widen` it is.